A38-admin-interface-api.md

```diff
@@ -0,0 +1,200 @@
+
+In addition, by providing new APIs, we abstract away the individual services with the concept admin services. We can also achieve **dynamic admin services selection**, for example, we **only add CSDS service if the xDS package is pulled in**. In this way, we won’t force users to take unnecessary dependencies.
+
+
+### Types of Services
+
+To help communicate the scenarios, we can define all services into several types, from the view of administrative functions:
+
+* **Primary service**: the user-provided services that handle the application logic;
+* **Binary-global service**: the services that expose the state of the entire gRPC library, and they serve the same information regardless of which server they have been registered (e.g., Channelz, CSDS);
+* **Per-server service**: the services that distribute the state of other registered services on the same server (e.g., Health, Reflection).
+
+The binary-global services contain sensitive information about traffic, configuration, even security, hence we hope users can set up the services securely in our recommended usage or the upcoming new API.
+
+The recommended way of securing the admin service port is binding it to either `localhost` or using a `UDS` port.
+
+
+## Proposal
+
+There are several different options for the convenience API design to help create a more usable admin interface API. Each option is a tradeoff between simplicity to use and flexibility. The simplicity to use is defined by lines of code to create a serving gRPC server with admin services, and the additional cognitive cost to learn the new API or manage dependencies.
+
+The flexibility is defined by how can users provide additional server creation options. To date, users can specify one or more **listening addresses, credentials, TCP/HTTP config, codec, interceptors and service implementations** during the server creation process. Not to mention the potential  impact on the threading model during the start of the server. If we want to bundle as many options as possible into one line of code, we will inevitably sacrifice some flexibility.
+
+Options to create an admin server are ordered from the most flexible to the least:
+
+1. The control group: using existing API to create the server;
+2. A function that combines admin services from different packages; **<-RECOMMENDED BY THIS PROPOSAL**
+3. A method of the gRPC server class that adds admin services;
+4. A function that creates the gRPC server object bound with admin services;
+5. A function that creates and starts the gRPC server with admin services.
+
+This proposal will recommend option 2 as the admin interface API (for other alternatives, see rationale section below). The usage snippet of admin interface API in each language:
+
+
+### Java
```

Comment:
I see Java, Go, and C++ here.  What about Python and the other wrapped languages?

Reply:
They are not yet on the road map, since CSDS is only going to support the 3 primary gRPC languages. In Python and other wrapper languages, it would a thin layer on top of Channelz, which wouldn't provide much value.

Reply:
I can understand if we decide that implementing CSDS in the other languages is a lower priority, but there's no reason why we can't or shouldn't eventually support it in all languages.  And the fact that we may want to add it later if exactly the justification for having an admin interface API: it means that we can add CSDS later without applications needing to change.

This seems like it should be something fairly simple to do in all languages, and doing it now will give us flexibility later.  Why would we *not* want to do this?  @srini100

Reply:
We are short on time and resources to consider wrapped languages at this time. It is probably better to add the admin APIs in wrapped languages when we can get functional parity with other languages (i.e., CSDS needs to be implemented first). Otherwise, users have to worry about the differences in admin services in different language.

Reply:
I disagree that it's better to wait.  The whole benefit of the admin API is that it allows us to add new admin services (like CSDS) automatically when they become available.  So even if we don't have CSDS implemented in wrapped languages yet, implementing the admin API now will make it easier to add CSDS later.

If we aren't going to implement this in wrapped languages right now because we don't have the resources to get the work done, then we can live with that.  But let's not pretend that that's the preferred outcome.